File: cpp/120_minimum_total/minimum_total.cpp
#include <algorithm>
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int minimumTotal(vector<vector<int>>& triangle) {
    int n = triangle.size();
    vector<int> f(n);
    f[0] = triangle[0][0];
    for (int i = 1; i < n; ++i) {
      f[i] = f[i - 1] + triangle[i][i];
      for (int j = i - 1; j > 0; --j) {
        f[j] = min(f[j - 1], f[j]) + triangle[i][j];
      }
      f[0] += triangle[i][0];
    }
    return *min_element(f.begin(), f.end());
  }
};

int main() {
  vector<tuple<vector<vector<int>>, int>> tests{
      {{{2}, {3, 4}, {6, 5, 7}, {4, 1, 8, 3}}, 11},
      {{{-10}}, -10},
  };

  for (auto& [triangle, ans] : tests) {
    assert(Solution().minimumTotal(triangle) == ans);
  }
}

File: rust/120_minimum_total/minimum_total.rs
struct Solution;

impl Solution {
    pub fn minimum_total(triangle: Vec<Vec<i32>>) -> i32 {
        let n = triangle.len();
        let mut f = vec![0; n];
        f[0] = triangle[0][0];

        for i in 1..n {
            f[i] = f[i - 1] + triangle[i][i];
            for j in (1..i).rev() {
                f[j] = std::cmp::min(f[j - 1], f[j]) + triangle[i][j];
            }
            f[0] += triangle[i][0];
        }

        *f.iter().min().unwrap()
    }
}

fn main() {
    let tests = vec![
        (
            vec![vec![2], vec![3, 4], vec![6, 5, 7], vec![4, 1, 8, 3]],
            11,
        ),
        (vec![vec![-10]], -10),
    ];

    for (triangle, ans) in tests {
        assert_eq!(Solution::minimum_total(triangle), ans);
    }
}
